Per-overload torch.ops API

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67254

Fixes https://github.com/pytorch/pytorch/issues/65997

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
Follow up work:
1. disallow `default` as an overload name for aten operators.
2. Add a method to obtain a list of all overloads (exclude the ones registered by JIT)
3. Add methods/properties to `OpOverload` to access more schema information (types of input and output args etc)

cc @ezyang @gchanan

Differential Revision: D33469839

anjali411 committed:
Per-overload torch.ops API

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"




cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"




cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"




cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"




cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 commented:
@pytorchbot ciflow rerun -l ciflow/all

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

albanD commented on torch/jit/annotations.py:
```diff
@@ -106,6 +110,9 @@ def is_vararg(the_callable):
 
 def get_param_names(fn, n_args):
+    if isinstance(fn, OpOverloadPacket):
+        fn = fn.default
```
not op here?

anjali411 replied:
good catch! updated it

albanD commented on torch/_ops.py:
```diff
@@ -25,6 +24,79 @@ def dl_open_guard():
     if _SET_GLOBAL_FLAGS:
         sys.setdlopenflags(old_flags)
 
+class OpOverload:
```
This class is also the one that will expose the tags when they are added right?

anjali411 replied:
possibly

albanD commented on torch/_ops.py:
```diff
@@ -25,6 +24,79 @@ def dl_open_guard():
+    def __deepcopy__(self, memo=None):
+        return self
+
+    # def __repr__(self):
```
nit: debug print.
Also if we want to add one here, it should be `__str__` I think?

anjali411 replied:
yup added it

albanD commented on torch/_ops.py:
```diff
@@ -25,6 +24,79 @@ def dl_open_guard():
+        self.op_name = op_name
+        self.op = op
+
+    def __deepcopy__(self, memo=None):
```
I think this is worth a comment above the class that this this object is immutable and must be unique for a given op. Also it is the responsibility of the user to ensure proper caching is done to not duplicate this object.
For this reason, the deepcopy is a no-op.

albanD replied:
Similar for the class above for a specific op/overload

anjali411 replied:
yup agreed, will add a comment. we are already caching these objects btw!

albanD replied:
> we are already caching these objects btw!

Yes, but it is now a requirement, not just a nice to have.

albanD commented on torch/_ops.py:
```diff
@@ -25,6 +24,79 @@ def dl_open_guard():
+        # to ensure torch.ops.foo.bar() is still callable from JIT
+        # "" key handling from above
+        # save a fn_ptr and just call that here. in its current version, it will be pretty slow
+        if kwargs is None:
```
You can inline this below with `**kwargs or {}`

albanD commented on torch/_ops.py:
```diff
@@ -25,6 +24,79 @@ def dl_open_guard():
+# second one is done by JIT. Creates a stack of all the overloads and then tries to match the correct one at runtime
+# and always calls into boxed version of the method
+
+# autograd codegen mostly creates variabletype, tracertype, inplace or view type and python bindings
```
Are these comments meant to remain?

anjali411 replied:
yeah I think we should keep these comments for future reference

albanD replied:
Ok, I think they should be cleaned up then, they are hard to read right now.

anjali411 replied:
cleaned up!

albanD reviewed (commented):
This will need to be looked at by someone from the jit team right?

anjali411 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1261,6 +1261,44 @@ void initJITBindings(PyObject* module) {
+          }
+          throw std::runtime_error("Found no matching schema");
+        } catch (const c10::Error& error) {
+          throw std::runtime_error(error.what_without_backtrace());
```
I followed the pattern in `_jit_get_operation` but I think you are right. I'll update both!

anjali411 commented on test/jit/test_custom_operators.py:
```diff
@@ -50,10 +50,6 @@ def test_default_arguments_are_used(self):
         output = torch.ops._test.leaky_relu(torch.tensor([-1.0, 1.0]))
         self.assertEqual(output, torch.tensor([-0.01, 1]))
 
-    def test_only_kwargs(self):
```
`self` can't be used as a kwarg anymore. read more here: https://github.com/pytorch/pytorch/pull/67254#issue-1036364695

anjali411 commented on torch/ao/quantization/quantization_mappings.py:
```diff
@@ -128,11 +128,11 @@
-    F.instance_norm: torch._ops.ops.quantized.instance_norm,
-    F.layer_norm: torch._ops.ops.quantized.layer_norm,
-    F.leaky_relu: torch._ops.ops.quantized.leaky_relu,
+    F.elu: torch.ops.quantized.elu,
```
I added a comment before but can't see it for some reason. This is not a necessary change but I don't see a reason why we need to get the ops from `torch._ops.ops` instead of `torch.ops`. I can undo it if there are other implications of this change

albanD replied:
Ho ok. To avoid confusion, these unrelated refactor can be done in a separate PR as well.

anjali411 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1261,6 +1261,44 @@ void initJITBindings(PyObject* module) {
+        return op->schema();
+      }
+    }
+    throw std::runtime_error("Found no matching schema");
```
Yeah I think so! This function is called from `getattr`, so I have `try catch` around the code and throw an `AttributeError`, and as a result users will never see this error.

anjali411 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1261,6 +1261,44 @@ void initJITBindings(PyObject* module) {
+          for (const auto& op: operations) {
+            if (op->schema().overload_name() == overload_name) {
+              std::ostringstream docstring;
+              docstring << "Automatically bound operator '" << op_name
```
removed it

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 commented on torch/_ops.py:
```diff
@@ -25,6 +25,98 @@ def dl_open_guard():
+
+        try:
+            use_key = "" if key == 'default' else key
+            op_ = torch._C._get_operation_overload(self.qualified_op_name, use_key)
```
TODO: disallow access to overloads registered by JIT. cc. @eellison is there a way to find out which overloads/ops are registered by JIT

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33262228).

gmagogsfm commented on torch/_ops.py:
```diff
@@ -25,6 +25,115 @@ def dl_open_guard():
+            except AttributeError:
+                throw_error = True
+
+        # We throw an error here as opposed to above since the stack trace and error message
+        # printed in that case is a bit confusing:
+        # RuntimeError: Found no matching schema
+        # During handling of the above exception, another exception occurred:
+        # AttributeError ...
+        if throw_error:
+            raise AttributeError("'{}' object has no attribute '{}'".format(str(self), key))
```
I think you can set the exception context to get rid of this logic. See more details here:
https://stackoverflow.com/a/52725410

anjali411 replied:
sweet! thanks for the suggestion -- that works

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -1270,7 +1270,7 @@ def forward(self, a) -> MyModule:
     if isinstance(obj, list):
         return create_script_list(obj)
-
+    isoverloadpacket = isinstance(obj, torch._ops.OpOverloadPacket)
```
Why is this logic needed? Is there plan to directly script an (overloaded) op?

anjali411 replied:
I looked up all instances in TorchScript where it was being checked if an object is a function or a method and added handling for OverloadPacket (to do the same thing you'd do if you received an op as before), just as a precaution. I can remove it if you think it's not needed

gmagogsfm replied:
Yeah, I don't think this would be needed, Let's try remove it and also add some test cases where per-overload op is used in TorchScript to see if everything still works.

gmagogsfm commented on test/test_per_overload_api.py:
```diff
@@ -0,0 +1,64 @@
+from torch.utils._python_dispatch import enable_python_mode
+import copy
+
+class TestPerOverloadAPI(TestCase):
```
Could we test that per-overload op is usable in a TorchScripted function?

anjali411 replied:
No fails with:
`RuntimeError: attribute lookup is not defined on builtin`
probably because TorchScript allows attribute lookups only on modules and OpOverloadPacket is not a module.

anjali411 replied:
full error log:

```
  File "/fsx/users/chourdiaanjali/tree1/test/test_per_overload.py", line 22, in test_torch_aten_add
    self.checkScript(fn, (torch.randn(2), torch.randn(2)))
  File "/fsx/users/chourdiaanjali/tree1/torch/testing/_internal/jit_utils.py", line 482, in checkScript
    self.checkScript(
  File "/fsx/users/chourdiaanjali/tree1/torch/testing/_internal/jit_utils.py", line 466, in checkScript
    cu = torch.jit.CompilationUnit(script, _frames_up=frames_up)
RuntimeError:
attribute lookup is not defined on builtin:
  File "<string>", line 2
def fn(a, b):
    return torch.ops.aten.add.Tensor(a, b)
```

gmagogsfm replied:
Is per-overload op intended an eager-only feature?

anjali411 replied:
yeah

gmagogsfm replied:
It might be a little confusing for users if `torch.ops.namespace.op_a` works but `torch.ops.namespace.op_a.overload` doesn't.  What do you think?

anjali411 replied:
We intend to use the `OpOverload` API in torch_dispatch, so it comes down to whether or not that's composable with TorchScript. In the current state, I think it would be ok to not add the TorchScript support.

gmagogsfm replied:
Alright, if it is not intended for end users we can skip TorchScript support. What do you think about improving the error message? Maybe we should add a special case there to let users know that per-overload op invocation isn't supported in TorchScript? The current message isn't clear or actionable enough for them. What do you think?

anjali411 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1261,6 +1261,44 @@ void initJITBindings(PyObject* module) {
+      }
+      throw std::runtime_error("Found no matching schema");
+    } catch (const c10::Error& e) {
+      auto msg = torch::get_cpp_stacktraces_enabled() ? e.what() : e.what_without_backtrace();
+      throw std::runtime_error(msg);
```
I followed error handling semantics in `_jit_get_operation`. would you suggest removing the try catch here?

albanD replied:
If you remove the try-catch, you will need to use the `HANDLE_TH_ERRORS` macros I think (that is the way we do the maping usually).

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

TODO: disallow `default` as an overload name for aten operators.

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
cc ezyang gchanan

Differential Revision: [D33262228](https://our.internmc.facebook.com/intern/diff/D33262228)

[ghstack-poisoned]

anjali411 commented on torch/_ops.py:
```diff
@@ -25,6 +25,115 @@ def dl_open_guard():
+
+    # TODO: add more methods to expose information about input and output arguments
+
+class OpOverloadPacket():
```
done!

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33262228).

albanD commented on test/test_per_overload_api.py:
```diff
@@ -0,0 +1,61 @@
+
+class TestPerOverloadAPI(TestCase):
+    def test_basics_opoverloadpacket(self):
+        add_packet = torch.ops.aten.add
```
Could you add a comment here mentioning that add is only used as an example here and it is ok to upadate this test if the semantic of add is modified in the future (same for the test below).

albanD commented on torch/_ops.py:
```diff
@@ -25,6 +25,109 @@ def dl_open_guard():
+        return self.__schema.overload_name
+
+    @property
+    def overloadpacket(self):
```
`overload_packet` would be more consistent with the other property above no?

albanD commented on torch/_ops.py:
```diff
@@ -25,6 +25,109 @@ def dl_open_guard():
+# You can obtain an OpOverload object through attribute query on OpOverloadPacket.
+class OpOverload:
+    def __init__(self, overloadpacket, op, schema):
+        self.__op = op
```
Why double _?

albanD reviewed (commented):
Looks quite good. I think we'll want to discuss the `is` / `in` / `==` semantic for these but it can be done in a follow up.

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
Follow up work:
1. disallow `default` as an overload name for aten operators.
2. Add a method to obtain a list of all overloads (exclude the ones registered by JIT)
3. Add methods/properties to `OpOverload` to access more schema information (types of input and output args etc)

cc ezyang gchanan

Differential Revision: [D33262228](https://our.internmc.facebook.com/intern/diff/D33262228)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33262228).

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
Follow up work:
1. disallow `default` as an overload name for aten operators.
2. Add a method to obtain a list of all overloads (exclude the ones registered by JIT)
3. Add methods/properties to `OpOverload` to access more schema information (types of input and output args etc)

cc ezyang gchanan

Differential Revision: [D33262228](https://our.internmc.facebook.com/intern/diff/D33262228)

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
Follow up work:
1. disallow `default` as an overload name for aten operators.
2. Add a method to obtain a list of all overloads (exclude the ones registered by JIT)
3. Add methods/properties to `OpOverload` to access more schema information (types of input and output args etc)

cc ezyang gchanan

Differential Revision: [D33262228](https://our.internmc.facebook.com/intern/diff/D33262228)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33262228).

albanD commented on torch/_ops.py:
```diff
@@ -25,6 +25,109 @@ def dl_open_guard():
+        self._schema = schema
+        self._overloadpacket = overloadpacket
+
+    # it's a no-op since OpOverload object is immutable and must be for a given op overload.
```
missing "unique"?
(same for the one below)

anjali411 replied:
I'll add it as a TODO for the follow-up PR. don't really wanna deal with exporting the internal changes:)

albanD reviewed (approved):
nit on doc. but LGTM.

ezyang commented on torch/_ops.py:
```diff
@@ -25,6 +25,109 @@ def dl_open_guard():
+            schema = torch._C._get_schema(self._qualified_op_name, use_key)
+            overload = OpOverload(self, op_, schema)
+            # cache the overload object
+            setattr(self, key, overload)
```
This makes me nervous. OpOverload contains a pointer to the OpOverloadPacket (self) that it was constructed from. This looks like a reference cycle to me. Luckily it doesn't matter too much, because ops don't really get GC'ed, but it gives me bad vibes.

anjali411 replied:
is the issue here that we are giving up on instant deletion because of the ref cycle? would you suggest not having a pointer from the opoverload to opoverloadpacket?

suo commented:
It looks like this PR broke ONNX tests on master? https://www.torch-ci.com/minihud?name_filter=linux-xenial-py3.7-clang7-onnx%20/%20test%20(default,%201,%202,%20linux.2xlarge). This commit appeared to succeed, but the that's because underlying build failed so the test job never ran. Given that this PR also has the same error, I am reverting.

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
Follow up work:
1. disallow `default` as an overload name for aten operators.
2. Add a method to obtain a list of all overloads (exclude the ones registered by JIT)
3. Add methods/properties to `OpOverload` to access more schema information (types of input and output args etc)

cc ezyang gchanan

Differential Revision: [D33262228](https://our.internmc.facebook.com/intern/diff/D33262228)

[ghstack-poisoned]

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
Follow up work:
1. disallow `default` as an overload name for aten operators.
2. Add a method to obtain a list of all overloads (exclude the ones registered by JIT)
3. Add methods/properties to `OpOverload` to access more schema information (types of input and output args etc)

cc ezyang gchanan

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33469839).

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
Follow up work:
1. disallow `default` as an overload name for aten operators.
2. Add a method to obtain a list of all overloads (exclude the ones registered by JIT)
3. Add methods/properties to `OpOverload` to access more schema information (types of input and output args etc)

cc ezyang gchanan

Differential Revision: [D33469839](https://our.internmc.facebook.com/intern/diff/D33469839)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33469839).

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
Follow up work:
1. disallow `default` as an overload name for aten operators.
2. Add a method to obtain a list of all overloads (exclude the ones registered by JIT)
3. Add methods/properties to `OpOverload` to access more schema information (types of input and output args etc)

cc ezyang gchanan

Differential Revision: [D33469839](https://our.internmc.facebook.com/intern/diff/D33469839)

[ghstack-poisoned]

anjali411 commented:
`@pytorchbot ciflow rerun -l ciflow/all

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33469839).

anjali411 committed:
Update on "Per-overload torch.ops API"


Fixes https://github.com/pytorch/pytorch/issues/65997

BC breaking:
`output = torch.ops._test.leaky_relu(self=torch.tensor(-1.0))` now fails with the error `TypeError: __call__() got multiple values for argument 'self'` since we call into `OpOverloadBundle`'s `__call__` method that has `self` bound to it as its first argument.
            
Follow up work:
1. disallow `default` as an overload name for aten operators.
2. Add a method to obtain a list of all overloads (exclude the ones registered by JIT)
3. Add methods/properties to `OpOverload` to access more schema information (types of input and output args etc)

cc ezyang gchanan

Differential Revision: [D33469839](https://our.internmc.facebook.com/intern/diff/D33469839)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33469839).
